### ISSUE TITLE
Out Of Memory Error After Some Number of Iterations

### ISSUE THREAD

**Issue by errissa** [mitsuba-renderer/mitsuba3#433]
## Summary

I get the following error after some number of iterations of a gradient-based optimization loop:

`RuntimeError: jit_malloc(): out of memory! Could not allocate 268435456 bytes of device memory.`

## System configuration

System information:

  OS: Ubuntu 20.04.5 LTS
  CPU: 12th Gen Intel(R) Core(TM) i9-12900H
  GPU: NVIDIA GeForce RTX 3080 Ti Laptop GPU
  Python: 3.8.10 (default, Jun 22 2022, 20:18:18) [GCC 9.4.0]
  NVidia driver: 515.65.01
  CUDA: 11.7.64
  LLVM: 12.0.0

  Dr.Jit: 0.3.2
  Mitsuba: 3.1.1
     Is custom build? False
     Compiled with: GNU 10.2.1
     Variants:
        scalar_rgb
        scalar_spectral
        cuda_ad_rgb
        llvm_ad_rgb

## Description

I have implemented a gradient-based optimization of some texture maps using the "Gradient based optimization" tutorial. It follows the same basic loop of rendering an image, computing loss, doing back propagation, stepping the optimizer, then updating scene parameters. With each iteration the amount of GPU memory consumed increases a little bit and after some number of iterations (>300 with my test scene), GPU memory runs out and I receive the following error:
```
Traceback (most recent call last):
  File "scripts/mi_scene_estimation.py", line 226, in <module>
    run_estimation('-continue' in sys.argv, int(sys.argv[1]), '-last' in sys.argv)
  File "scripts/mi_scene_estimation.py", line 175, in run_estimation
    dr.backward(loss)
  File "/home/renes/development/intel_work/o3dpy/lib/python3.8/site-packages/drjit/router.py", line 4523, in backward
    backward_from(arg, flags)
  File "/home/renes/development/intel_work/o3dpy/lib/python3.8/site-packages/drjit/router.py", line 4469, in backward_from
    traverse(ta, _dr.ADMode.Backward, flags)
  File "/home/renes/development/intel_work/o3dpy/lib/python3.8/site-packages/drjit/router.py", line 4350, in traverse
    dtype.traverse_(mode, flags)
  File "/home/renes/development/intel_work/o3dpy/lib/python3.8/site-packages/mitsuba/python/util.py", line 400, in backward
    self.integrator.render_backward(self.scene, self.params, self.grad_out(),
  File "/home/renes/development/intel_work/o3dpy/lib/python3.8/site-packages/mitsuba/python/ad/integrators/common.py", line 938, in render_backward
    image = sensor.film().develop()
RuntimeError: jit_malloc(): out of memory! Could not allocate 268435456 bytes of device memory.
drjit-autodiff: variable leak detected (154 variables remain in use)!
 - variable a1135830 (1 references)
 - variable a1135726 (2 references)
 - variable a1135739 (16 references)
 - variable a1135694 (3 references)
 - variable a1135741 (1 references)
 - variable a1135777 (1 references)
 - variable a1135725 (2 references)
 - variable a1135758 (1 references)
 - variable a1135795 (1 references)
 - variable a1135692 (2 references)
 - (skipping the rest)
drjit-autodiff: edge leak detected (214 edges remain in use)!
```
Does DrJIT require more memory with each iteration and therefore this behavior is expected?

**Comment by Speierers** [comment]
Hi @errissa ,

This isn't the expected behavior. Could you please share with us you Python script to reproduce this issue so we can take a look?

**Comment by wjakob** [comment]
Closing due to a lack of response.

**Comment by errissa** [comment]
@Speierers Sorry for never responding. I got busy with other work and worked around the memory leak. I did recently find the cause so I'm documenting it here in case others have the same issue. I had created a `scene` with the `path` integrator but inside the optimization loop I had this:
```
img = mi.render(scene, params, spp=8, seed=i, integrator=mi.load_dict({'type': 'prb'}))
```
Creating a new integrator with each loop apparently causes a GPU memory leak. I moved `mi.load_dict(...)` outside the loop thereby creating it only once and changed the rendering to:
```
img = mi.render(scene, params, spp=8, seed=i, integrator=prb_integrator)
```
No GPU memory leak!
